### PR TITLE
Added a callback to stringify and parse

### PR DESCRIPTION
**Pull request by wbyoung**
This allows much flexibility including allowing users to support custom types.

### CODE REVIEW

**Comment by hunterloftis** [comment]
Good PR, although the design goal of cryo is to be a lib that 'just works.' We should consider implementing a robust constructor restoration as just a normal part of stringify and parse instead of something that the user has to specify themselves.

**Comment by wbyoung** [comment]
@hunterloftis the problem is knowing where to find classes. In many styles, the class won't even be accessible to `cryo`.

Namespacing make classes unavailable:

```js
window.MyApp.Person = function Person() {
};
```

IIFEs make classes unavailable:

```js
(function() {
  function Person() {
  }

  cryo.prase(...)

}());
```

AMD modules make classes unavailable:

```js
define('person', ['module'], function(module) {
  function Person() {}
  module.exports = Person;
});

cryo.parse(...);
```

The only case that `cryo` could handle in a "just works" fashion would be when classes are defined on `window` (global), but that wouldn't work in Node.js either.

Also, there's no way to tell whether or not classes will actually be defined with a name or defined anonymously:

```js
var Person = function() {
};
```

Basically, the callback would be really useful regardless of how it's used. Otherwise, lib authors have to write a traversal pre & post serialization as well. I'm sure there's a lib to handle it, but the callback is a nice convenience.

**Comment by hunterloftis** [comment]
Good points. Thanks for the PR with tests, merging!

**Comment by wbyoung** [comment]
Thanks for accepting it and for creating the project! :smile:
